File: examples/exo/cpp/src/handset.cpp
#include "exo_example_support.hpp"

#include <cstdlib>
#include <exception>
#include <iomanip>
#include <iostream>
#include <string>

namespace {

void print_handsets(const qnbot::Sample<qnbot::ExoTelemetry>& sample) {
    const auto& payload = sample.value.payload;
    std::cout << "handsets sequence=" << sample.sequence << " left=("
              << payload.left_handset.axis_x_raw << ", "
              << payload.left_handset.axis_y_raw
              << ", trigger=" << payload.left_handset.trigger_raw
              << ", buttons=0x" << std::hex << payload.left_handset.button_mask
              << std::dec << ") right=(" << payload.right_handset.axis_x_raw
              << ", " << payload.right_handset.axis_y_raw
              << ", trigger=" << payload.right_handset.trigger_raw
              << ", buttons=0x" << std::hex << payload.right_handset.button_mask
              << std::dec << ")\n";
}

} // namespace

int main(int argc, char** argv) {
    try {
        std::string port;
        std::string name = "primary";
        for (int index = 1; index < argc; ++index) {
            const std::string argument = argv[index];
            if (argument == "--port") {
                port = example::require_value(argc, argv, index, argument);
            } else if (argument == "--name") {
                name = example::require_value(argc, argv, index, argument);
            } else {
                throw std::invalid_argument("unknown argument: " + argument);
            }
        }
        if (port.empty()) throw std::invalid_argument("--port is required");

        qnbot::Sdk sdk(example::serial_config(port, name));
        auto exo = sdk.exo();
        auto device = exo.device(name);
        if (!device.get_device_info().capabilities.handsets) {
            std::cout << "this device reports no handset" << std::endl;
            exo.close();
            return EXIT_SUCCESS;
        }

        auto telemetry = device.telemetry();
        const auto subscription = telemetry.subscribe(print_handsets);

        exo.start();
        std::cout << "running; press Ctrl+C to stop" << std::endl;
        exo.run_forever();
        static_cast<void>(subscription);
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}

File: exo/python/src/handset.py
from __future__ import annotations

import argparse

from qnbot_sdk import Sample, Sdk, SerialConnection
from qnbot_sdk.exo import ExoConfig, ExoTelemetry


def create_sdk(port: str, name: str) -> Sdk:
    return Sdk(devices=(ExoConfig(name=name, connection=SerialConnection(port=port)),))


def print_handsets(sample: Sample[ExoTelemetry]) -> None:
    payload = sample.value.payload
    print(
        f"handsets sequence={sample.sequence} "
        f"left=({payload.left_handset.axis_x_raw}, "
        f"{payload.left_handset.axis_y_raw}, "
        f"trigger={payload.left_handset.trigger_raw}, "
        f"buttons=0x{payload.left_handset.button_mask:04x}) "
        f"right=({payload.right_handset.axis_x_raw}, "
        f"{payload.right_handset.axis_y_raw}, "
        f"trigger={payload.right_handset.trigger_raw}, "
        f"buttons=0x{payload.right_handset.button_mask:04x})"
    )


def main() -> None:
    arguments = argparse.ArgumentParser(description="Read Exo handset states")
    arguments.add_argument("--port", required=True, help="Serial port for the device")
    arguments.add_argument("--name", default="primary", help="Logical device name")
    options = arguments.parse_args()

    sdk = create_sdk(options.port, options.name)
    exo = sdk.exo()
    device = exo.device(options.name)
    if not device.get_device_info().capabilities.handsets:
        print("this device reports no handset", flush=True)
        exo.close()
        return

    device.telemetry().subscribe(print_handsets)
    exo.start()
    print("running; press Ctrl+C to stop", flush=True)
    exo.run_forever()


if __name__ == "__main__":
    main()
